### PR TITLE
FIX Make eignvectors consistent

### PR DESCRIPTION
**Pull request by thomasjpfan**
Here is a quick fix to make the eigenvectors consistent.

### CODE REVIEW

File: sklearn/decomposition/tests/test_fastica.py [hunk 1]
```diff
@@ -399,3 +387,38 @@ def test_fastica_output_shape(whiten, return_X_mean, return_n_iter):
     assert len(out) == expected_len
     if not whiten:
         assert out[0] is None
+
+
+@pytest.mark.parametrize("add_noise", [True, False])
+@pytest.mark.parametrize("seed", range(2))
+def test_fastica_simple_different_solvers(add_noise, seed):
```
Comment (thomasjpfan) [inline sklearn/decomposition/tests/test_fastica.py:394]: This is a new test to check for consistency based on what you had.

File: sklearn/decomposition/tests/test_fastica.py [hunk 2]
```diff
@@ -119,36 +119,24 @@ def g_test(x):
...
-    for solver in ("eigh", "svd"):
-        _, _, sources_fun = fastica(
-            m.T, fun=nl, algorithm=algo, random_state=0, svd_solver=solver
-        )
-        ica = FastICA(fun=nl, algorithm=algo, random_state=0, svd_solver=solver)
-        sources = ica.fit_transform(m.T)
-        outs[solver] = sources
-        assert ica.components_.shape == (2, 2)
-        assert sources.shape == (1000, 2)
+    _, _, sources_fun = fastica(m.T, fun=nl, algorithm=algo, random_state=seed)
```
Comment (thomasjpfan) [inline sklearn/decomposition/tests/test_fastica.py:122]: This reverts the tests to what they were on `upstream/main`.

File: sklearn/decomposition/_fastica.py [hunk 3]
```diff
@@ -574,12 +574,13 @@ def g(x, fun_args):
                     )
                 d[degenerate_idx] = eps  # For numerical issues
                 d = np.sqrt(d, d)
-                d, u = d[sort_indices], u[sort_indices]
-                # Resize and reorder to match svd
-                u = u[::-1, : min(X.shape) : -1]
+                d, u = d[sort_indices], u[:, sort_indices]
```
Comment (thomasjpfan) [inline sklearn/decomposition/_fastica.py:577]: Based on [eigh](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigh.html) docs, the eigenvectors has shape `(M, N)` where `N` is the number of eigenvalues. Thus, I think the sorting should be done over `axis=1`.

**Comment by thomasjpfan** [comment]
<s>After thinking it over, I do not think this is correct.</s>

Nevermind I think this works.

**Comment by thomasjpfan** [comment]
Here is a code snippet to demonstrate:

```python
from scipy import linalg
import numpy as np
from numpy.random import default_rng
from numpy.testing import assert_allclose

def my_svd(X):
    u, d = linalg.svd(X, full_matrices=False)[:2]
    u *= np.sign(u[0])
    return u, d

def my_eigh(X):
    d, u = linalg.eigh(X.T.dot(X))
    sort_indicies = np.argsort(d)[::-1]
    d = d[sort_indicies]
    u = u[:, sort_indicies]
    u *= np.sign(u[0])
    return u, np.sqrt(d)

rng = default_rng()
for i in range(10):
    m, n = 9, 6
    X = rng.standard_normal((m, n))
    u_svd, d_svd = my_svd(X.T)
    u_eigen, d_eigen = my_eigh(X)
    
    assert_allclose(d_svd, d_eigen)
    assert_allclose(u_svd, u_eigen)
```

**Comment by thomasjpfan** [comment]
Although, we can use the align the eigenvectors based on the "maximum":

```python
from scipy import linalg
import numpy as np
from numpy.random import default_rng
from numpy.testing import assert_allclose

def my_svd(X):
    u, d = linalg.svd(X, full_matrices=False)[:2]
    max_abs_cols = np.argmax(np.abs(u), axis=0)
    signs = np.sign(u[max_abs_cols, range(u.shape[1])])
    u *= signs
    
    return u, d

def my_eigh(X):
    d, u = linalg.eigh(X.T.dot(X))
    sort_indicies = np.argsort(d)[::-1]
    d = d[sort_indicies]
    u = u[:, sort_indicies]
    max_abs_cols = np.argmax(np.abs(u), axis=0)
    signs = np.sign(u[max_abs_cols, range(u.shape[1])])
    u *= signs
    
    return u, np.sqrt(d)

rng = default_rng()
for i in range(10):
    m, n = 9, 6
    X = rng.standard_normal((m, n))
    u_svd, d_svd = my_svd(X.T)
    u_eigen, d_eigen = my_eigh(X)
    
    assert_allclose(d_svd, d_eigen)
    assert_allclose(u_svd, u_eigen)
```

File: sklearn/decomposition/_fastica.py [hunk 4]
```diff
@@ -574,12 +574,13 @@ def g(x, fun_args):
...
             else:
                 u, d = linalg.svd(XT, full_matrices=False, check_finite=False)[:2]
 
+            # Give consistent eigenvectors for both svd solvers
+            u *= np.sign(u[0])
```
Comment (thomasjpfan) [inline sklearn/decomposition/_fastica.py:582]: The alternative solution is:

```python
    max_abs_cols = np.argmax(np.abs(u), axis=0)
    signs = np.sign(u[max_abs_cols, range(u.shape[1])])
    u *= signs
```

If we want to be strictly the same as `svd_flip`.
